### PR TITLE
Make required actions configurable (#28400)

### LINKED ISSUE #28400: Make RequiredActions configurable
### Description

This feature request asks to add support for configurable properties to RequiredActions. 
Currently support for configuration of required actions via the admin ui are quite limited. 
The only allowed configuration options are: 
- enabled (is the required action enabled at all?)
- default (is the required action set for new users by default?) 
- order / priority (controls the order in which required actions is presented to the user)

It would be helpful to allow developers to specify additional configurable properties `org.keycloak.provider.ProviderConfigProperty` for required actions, similar to how authenticator executions can be configured.

### Discussion

_No response_

### Motivation

Behaviour for required actions should be configurable. Currently one has to pull configuration from realm attributes, env variables or other means.
Having official support for configuration would allow us to have a common way to configure actions. 

### Details

# Implementation ideas

## Configuration options 

There are multiple ways to implement support for configuration options for required actions, here are 2 options:

### Option 1
Leverage `org.keycloak.provider.ProviderFactory#getConfigMetadata()`
Advantages:
- easier, as it is already available API

Disadvantages:
- no default config 
- no help text 

### Option 2
Make required actions infrastructure aware of `org.keycloak.provider.ConfiguredProvider` 
Advantages:
- more options (help text, default config) 
- Developers could opt-in to make their required actions configurable by implementing the `ConfiguredProvider` interface.

Disadvantages:
- a bit more work
- 2 different ways to define config properties (`ProviderFactory#getConfigMetadata()` vs. `ConfiguredProvider#getConfigProperties()`) 

I currently lean toward Option 1.

## Required Action Configuration storage

The required actions infrastructure already supports the persistence of configuration metadata via 
[`org.keycloak.models.jpa.entities.RequiredActionProviderEntity#config`](https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/entities/RequiredActionProviderEntity.java#L76) in the `REQUIRED_ACTION_CONFIG` table.

## REST API extensions in AuthenticationManagementResource 

The `AuthenticationManagementResource` already supports the management of required actions. It seems as the natural place to add support for required action config management.

I propose to add the following new endpoints to manage required action configuration within a realm bound to the required action "alias" / provider id.
```
  @Path("required-actions/{alias}/config-description")
  @GET
  @Produces(MediaType.APPLICATION_JSON)
  RequiredActionConfigInfoRepresentation getRequiredActionConfigDescription(@PathParam("alias") String alias);

  @Path("required-actions/{alias}/config")
  @GET
  @Produces(MediaType.APPLICATION_JSON)
  RequiredActionConfigRepresentation getRequiredActionConfig(@PathParam("alias") String alias);

  @Path("required-actions/{alias}/config")
  @DELETE
  void removeRequiredActionConfig(@PathParam("alias") String alias);

  @Path("required-actions/{alias}/config")
  @PUT
  @Consumes(MediaType.APPLICATION_JSON)
  void updateRequiredActionConfig(@PathParam("alias") String alias, RequiredActionConfigRepresentation rep);
```

## Additional work
- Introduce RequiredActionConfigModel
- Extend RequiredActionContext to expose RequiredActionConfigModel
- Extend admin-client libraries (jakarta, javaee, js/ts)
- Extend admin ui (required action list with config button and required action config form)
- Update documentation
- Update tests (REST API, UI tests)

### PR DESCRIPTION
**Pull request by thomasdarimont**
This PR adds support for configurable required actions.

Some implementation notes:

- Introduced `RequiredActionConfigModel` to hold the configration of a RequiredAction
- Extended `AuthenticationManagementResource` with CRUD operations for RequiredAction config management
- Extended admin-client libraries with RequiredAction config management methods
- Exposed `RequiredActionConfigModel` via RequiredActionContext to users
- Add tests for crud operations on configurable required actions

Fixes #28400

### CODE REVIEW

**Comment by thomasdarimont** [comment]
@stianst this is the PR for initial support for configurable required actions.

What's already working:
- Storage of required actions configuration in the `required_action_config` table
- Configuration per required actions instance in a realm
- CRUD for required actions with configuration
- Users can make their own required action implementations configurable by overriding the `org.keycloak.provider.ProviderFactory#getConfigMetadata()`
- Initial tests for RequiredAction CRUD config management
- Extended admin-client libraries

What's still missing:
- Admin UI integration ("cog" settings button and settings form for required actions)
- Documentation
- "Proper" way to validate config settings before save

**Review by thomasdarimont — COMMENTED** [review]
This is just a placeholder for the configure button

File: js/apps/admin-ui/src/authentication/RequiredActions.tsx
Comment (thomasdarimont) [inline js/apps/admin-ui/src/authentication/RequiredActions.tsx:140]: This is just a placeholder for the "configure" button for required actions.

File: server-spi-private/src/main/java/org/keycloak/authentication/RequiredActionFactory.java
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/authentication/RequiredActionFactory.java:65]: I'm not sure whether this is the proper way to perform validation, but it could be used to prevent bad configurations to be written to the database.

File: server-spi-private/src/main/java/org/keycloak/utils/CredentialHelper.java
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/utils/CredentialHelper.java:98]: This and the following helper functions should be moved elsewhere... just placed them here because the authenticator config methods were also located here.

**Review by thomasdarimont — COMMENTED** [review]
Some remarks inline.

**Comment by thomasdarimont** [comment]
I now have an initial version working, that allows to inspect, add, update and remove required action configuration via the admin ui.

Required Actions List:
I now only show the settings icon, if the required action is configurable. Instead of using a dedicated column for the button, we could also place the icon next to the required action label.
image

Required Actions Config Modal:
In the config modal I currently only show the actual settings. It would be good to show the actual name of the current required action as a label somewhere.
image

**Review by thomasdarimont — COMMENTED** [review]
Added some more remarks to the places where I'm unsure

**Comment by thomasdarimont** [comment]
Thanks to @edewit the new UI now looks a bit better:
image

Config screen:
image

@edewit would it be possible to show the translated name instead of the alias? (Update Password instead of "UPDATE_PASSWORD"). Also I think the config dialog would scale better if the name of the required action would be on a separat line. Some users tend to have quite long (but descriptive) names for required actions.

File: server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java
Comment (mposolda) [inline server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java:127]: Do we need `isConfigurable` and `setConfigurable` on the model? It seems that the `configurable` field is not backed in the DB and it is always derived based on the factory (like in `RealmAdapter.entityToModel(RequiredActionProviderEntity entity)`).

Also I believe it is not needed in the JSON representation object, which is used for export/import, but I understand it can be useful for have this in the UI layer. Is it possible to decouple those?

**Review by mposolda — COMMENTED** [review]
Thanks,

I've added one minor comment inline. In general, I am not 100% sure if we want to rely on `ComponentModel`, which was the approach discussed on the meeting AFAIR. But maybe not... (added some more details in the email).

File: server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java
Comment (thomasdarimont) [reply server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java:127]: Thanks for taking a look! Yes I mostly added the configurable property to the `RequiredActionProviderModel` and `RequiredActionProviderRepresentation` . Initially I thought about deriving the "configurability" of a RequiredAction from a non-empty configuration. However, with this we currently cannot differentiate an empty configuration from a non-configurable required action, as the config propery is currently always initialized with an empty map here: https://github.com/keycloak/keycloak/blob/main/core/src/main/java/org/keycloak/representations/idm/RequiredActionProviderRepresentation.java#L35
If we would change this to be `null` by default, then I could remove the `configurable` from the model and the representation property. 

I was also thinking about using the component model but then realized that required actions already have a database backed config HashMap, so I kept using that.

**Comment by thomasdarimont** [comment]
A quick experiment for better validation support: 
image

**Comment by thomasdarimont** [comment]
Now using the required action `name` instead of `alias`, as this is more user friendly.
image

File: server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java
Comment (mposolda) [reply server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java:127]: I understand that the info about "configurable" is needed in the UI. But I think it is not needed in the model layer and might be cleaner to remove it from here.

I see the options like:

1) Keep `configurable` at the model layer and `RequiredActionProviderRepresentation`. Not so clean IMO.
2) Use the `null` to indicate. It can work, but just afraid a bit of things like `NullPointerException` etc.
3) Add stuff inside admin UI java extension (module `rest/admin-ui-ext` in the workspace) and make admin console to use it. I can imagine something like this in the package `org.keycloak.admin.ui.rest.model` inside admin UI extension (the stuff in the module `rest`, which contains REST endpoints used solely by admin console):

```
public class ConfigurableRequiredActionProviderRepresentation extends RequiredActionProviderRepresentation {

    private boolean configurable;

    // Getter and setter for "configurable" here ...
}
```

And then make `org.keycloak.admin.ui.rest.AuthenticationManagementResource` to return required actions like those, so the admin console UI is able to figure if action is configurable or not. The disadvantage of the extension is additional code.

My vote is either (2) or (3). WDYT?

**Comment by thomasdarimont** [comment]
@mposolda thanks for your review and the helpful suggestions! Regarding your proposal for moving forward, I went with a combination of (2) and (3) as discussued above:
- (2) use null to indicate no required action config is set
- (3) add  admin ui specific endpoint  that returns the `configurable` property for a required action.

The latest changes of this PR include:
- Removed boolean configurable from RequiredActionProviderRepresentation and RequiredActionConfigModel classes
- Adapted UI and API to work with requiredaction provider alias instead of providerId for consistency
- Introduced admin-ui specific endpoint org.keycloak.admin.ui.rest.AuthenticationManagementResource#getRequiredActions for retrieving required actions with config metadata

Since all the endpoints that handle requiredactions work with the requiredaction `alias` value,
I adapted all places that work with requiredaction configs to use aliases as well.
When I have the `providerId`, then I also check if the `providerId` matches that of the given `RequiredActionConfigModel`.

Only difference is, which only has an alias available at the moment:
org.keycloak.models.jpa.RealmAdapter#getRequiredActionConfigByAlias

However, using the aliases all the time means that required action aliases within a realm must be unique - is this guaranteed somehow?

I also tried to have `max_auth_age` as a generic configuration parameter for all required actions, but I needed to rollback that change I wasn't sure whether this really makes sense for all required actions and wanted to discuss this first. 

As an example I adapted the `UPDATE_PASSWORD` required action to consider the configured value for `max_auth_age` with a preference for the password policy if set.

I think the PR - at least for the initial capacity is on the finish line. Perhaps @edewit could take a look again on the translation of error messages if a validation fails :) (I need some help to render the proper error message based on the server response).
image

**Comment by thomasdarimont** [comment]
When we allow configuration of required actions, it might make sense to add support for help texts that can describe the action and options.

I experimented with this a bit here:
image

Adding helpText support for required actions was a brief experiment - I'll leave it out from this PR and add it in a separate PR.

File: js/apps/admin-ui/src/authentication/RequiredActions.tsx
Comment (thomasdarimont) [inline js/apps/admin-ui/src/authentication/RequiredActions.tsx:93]: We need to omit the `configurable` property here. This is necessary since we use the configurable action representation in the admin UI with the `configurable` property, which currently does not exist on the required action representation used by the standard admin rest API.

**Comment by thomasdarimont** [comment]
@mposolda I squashed the commits into one and rebased on current main.

**Comment by ahus1** [comment]
@thomasdarimont - there is another conflict, and it needs rebasing.

**Comment by thomasdarimont** [comment]
@ahus1 I just rebased the code on latest main and resolved the conflicts.

**Comment by thomasdarimont** [comment]
@jonkoops This PR currently fails due to an unrelated test:
```
  Running:  realm_settings_user_profile_enabled.spec.ts                                     (7 of 7)

  User profile tabs
    Attributes sub tab tests
      ✓ Goes to create attribute page (4828ms)
      ✓ Completes new attribute form and performs cancel (2892ms)
      ✓ Completes new attribute form and performs submit (2619ms)
      ✓ Modifies existing attribute and performs save (3569ms)
      ✓ Adds and removes validator to/from existing attribute and performs save (3593ms)
    Attribute groups sub tab tests 
      (Attempt 1 of 4) Deletes an attributes group  <---- This fails
      (Attempt 2 of 4) Deletes an attributes group
      (Attempt 3 of 4) Deletes an attributes group
      1) Deletes an attributes group
```

I ran the cypress tests locally and there it fails too. I currently don't see a link to my changes. Do you have an idea what the problem is?

**Comment by jonkoops** [comment]
Not sure what is happening there, perhaps someone from the UI team can take a look. @edewit @hmlnarik

**Comment by hmlnarik** [comment]
Looks like a test interference. I have introduced #29507 and rerun the admin UI testsuite from `Generate test seed` on.

**Comment by thomasdarimont** [comment]
Unfortunately I'm blocked by the failing UI tests, which are unrelated to this PR. The same tests also fail on my machine locally when I run the latest Keycloak main (I tested this with cypress on electron, and chrome).
The ui test that currently fails is: `client_authorization_test.spec.ts`.

Any help would be highly appreciated :)

**Comment by edewit** [comment]
> Any help would be highly appreciated :)

Tests have been fixed on main, I have rebased your PR

File: server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java
Comment (mposolda) [inline server-spi/src/main/java/org/keycloak/models/RequiredActionProviderModel.java:48]: Nitpick: This field is not used and can be removed from this class (probably all the changes from this class can be removed from this PR)

File: server-spi-private/src/main/java/org/keycloak/utils/CredentialHelper.java
Comment (mposolda) [reply server-spi-private/src/main/java/org/keycloak/utils/CredentialHelper.java:98]: Nitpick: Can be ok to introduce new class like `RequiredActionHelper` and add those methods here as they don't have much to do with credentials...

File: services/src/main/java/org/keycloak/authentication/requiredactions/UpdatePassword.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/requiredactions/UpdatePassword.java:256]: Nitpick: I can imagine it can be useful to have some util method (maybe in the `RequiredActionsHelper` as suggested in other comments or in `KeycloakModelUtils`) like:
```
RequiredActionConfigModel getRequiredActionConfig(RealmModel realm, String requiredActionProviderId)
```
as I suppose the similar code will apply always when retrieving configurations from required action

File: services/src/main/java/org/keycloak/authentication/requiredactions/UpdatePassword.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/requiredactions/UpdatePassword.java:78]:
```suggestion
                .helpText("Configures the duration in seconds this action can be used after the last authentication before the user is required to re-authenticate. This parameter is used just in the context of AIA when the kc_action parameter is available in the request, which is for instance when user himself updates his password in the account console. When the 'Maximum Authentication Age' password policy is used in the realm, it's value has precedence over the value configured here.") //
```

Nitpick: just suggesting to add a bit more info to the helpText to clarify it is used just in the context of AIA and about the fact that password policy has preference.

**Review by mposolda — COMMENTED** [review]
@thomasdarimont Thanks for all the updates!

I've added few comments inline. It can be good to address them, but at the same time, I don't consider any of them as a blocker, so we can possibly do as a follow-up. Whatever you prefer.

Will need review from UI team as well.

**Review by mposolda — DISMISSED** [review]

**Comment by mposolda** [comment]
@thomasdarimont Since there is conflict and this PR will require changes, WDYT about adding my "nitpicks" (if they makes sense for you) when you work on resolving conflict? :-) Thanks!

**Review by edewit — DISMISSED** [review]

**Comment by thomasdarimont** [comment]
@mposolda Thanks for your review and the recommendations. I fixed all your nitpicks and rebased to latest main. 
I think the PR is now ready for merge.

**Review by mposolda — APPROVED** [review]
@thomasdarimont @edewit Thanks a lot for all the updates and review!

**Pull request merged**
